server/src/main/java/org/opensearch/index/engine/OpenSearchReaderManager.java

```diff
@@ -78,13 +87,48 @@ protected void decRef(OpenSearchDirectoryReader reference) throws IOException {
 
     @Override
     protected OpenSearchDirectoryReader refreshIfNeeded(OpenSearchDirectoryReader referenceToRefresh) throws IOException {
-        final OpenSearchDirectoryReader reader = (OpenSearchDirectoryReader) DirectoryReader.openIfChanged(referenceToRefresh);
+        // This is directly from Lucene's SegmentInfosSearcherManager...
+        List<LeafReader> subs;
+        if (referenceToRefresh == null) {
+            subs = null;
+        } else {
+            subs = new ArrayList<>();
+            for (LeafReaderContext ctx : referenceToRefresh.getDelegate().leaves()) {
+                subs.add(ctx.reader());
+            }
+        }
+
+        final OpenSearchDirectoryReader reader;
+        // If not using NRT repl.
+        if (currentInfos == null) {
+          reader = (OpenSearchDirectoryReader) DirectoryReader.openIfChanged(referenceToRefresh);
+        } else {
+            // Open a new reader, sharing any common segment readers with the old one:
+            DirectoryReader innerReader = StandardDirectoryReader.open(referenceToRefresh.directory(), currentInfos, null, null);
```

Comment:
We should eventually pass subs to this open call that will reuse any segment readers vs creating new ones.  I'm currently setting it to null here because in the first go we have readers associated with different SegmentInfos. This is because we wiped the old segment files during copy.  To mitigate we need to turn off the initial indexing on a replica when a doc is added done through existing peer recovery and instead rely on a segCopy.  Doing this blows up here:
```
      // Make a best effort to detect when the app illegally "rm -rf" their
      // index while a reader was open, and then called openIfChanged:
      if (oldReader != null && Arrays.equals(commitInfo.info.getId(), oldReader.getSegmentInfo().info.getId()) == false) {
        throw new IllegalStateException("same segment " + commitInfo.info.name + " has invalid doc count change; likely you are re-opening a reader after illegally removing index files yourself and building a new index in their place.  Use IndexWriter.deleteAll or open a new IndexWriter using OpenMode.CREATE instead");
      }
```